### ISSUE TITLE
Fix atom_diag's hilbert space filter functionality (required for f-shell calculations)

### ISSUE THREAD

**Issue by HugoStrand** [TRIQS/triqs#633]
Dear All,

This is a note for the future about `atom_diag` and it's hilbert space functionality that is currently not exposed on the header level. This functionality needs to be restored to (re)-enable f-shell calculations with `cthyb`.

The `atom_diag_worker` has default arguments `n_min` and `n_max` that can be used to limit the range of total occupations accounted for in the diagonalization. This is currently broken and attempting to use it breaks in the `complete()` function when trying to diagonalize the sectors that are empty. The eigen solver throws with `matrix == empty`.

I have a commit with a test that reproduces the issue here 
https://github.com/HugoStrand/triqs/commit/f75586c85b4c7b6184510b13716e4386d6a254a9
for future reference.

Thank you @krivenko for guiding me through the `atom_diag` code. 👍 

Best, Hugo

```c++
71: Test command: /Users/hugstr/dev/triqs/cbuild/test/triqs/atom_diag/f_shell
71: Test timeout computed to be: 10000000
71: [==========] Running 1 test from 1 test case.
71: [----------] Global test environment set-up.
71: [----------] 1 test from atom_diag_real
71: [ RUN      ] atom_diag_real.f_shell_kanamori
71: unknown file: Failure
71: C++ exception with description ".. Triqs runtime error at /Users/hugstr/dev/triqs/triqs/arrays/linalg/eigenelements.hpp : 96
71: 
71: eigenelements_worker : the matrix is empty : matrix =  empty   
71: .. Error occurred on node 0
71: .. C++ trace is : 
71:     frame #6: 0x000000010e679b4b libtriqs.dylib`void triqs::arrays::linalg::eigenelements_worker<double>::_prepare<triqs::arrays::matrix<double, void> >(triqs::arrays::matrix<double, void> const&) const + 203
71:     frame #7: 0x000000010e67958e libtriqs.dylib`std::__1::pair<triqs::arrays::array<double, 1, void>, triqs::arrays::matrix<double, void> > triqs::arrays::linalg::eigenelements_worker<double>::eigenelements<triqs::arrays::matrix<double, void> >(triqs::arrays::matrix<double, void>&) const + 62
71:     frame #8: 0x000000010e669da9 libtriqs.dylib`std::__1::pair<triqs::arrays::array<double, 1, void>, triqs::arrays::matrix<triqs::arrays::matrix<double, void>::value_type, void> > triqs::arrays::linalg::eigenelements<triqs::arrays::matrix<double, void> >(triqs::arrays::matrix<double, void> const&) + 297
71:     frame #9: 0x000000010e6646ae libtriqs.dylib`triqs::atom_diag::atom_diag_worker<false>::complete() + 2638
71:     frame #10: 0x000000010e662778 libtriqs.dylib`triqs::atom_diag::atom_diag_worker<false>::autopartition() + 5976
71:     frame #11: 0x000000010e603b71 libtriqs.dylib`triqs::atom_diag::atom_diag<false>::atom_diag(triqs::operators::many_body_operator_generic<double> const&, triqs::hilbert_space::fundamental_operator_set const&, int, int) + 561
71:     frame #12: 0x000000010e5dfd35 f_shell`atom_diag_real_f_shell_kanamori_Test::TestBody() + 565
71:     frame #13: 0x000000010e7a4378 libgtest.dylib`void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) + 72
71:     frame #14: 0x000000010e7a423d libgtest.dylib`testing::Test::Run() + 1197
71:     frame #15: 0x000000010e7a5e64 libgtest.dylib`testing::TestInfo::Run() + 1188
71:     frame #16: 0x000000010e7a6787 libgtest.dylib`testing::TestCase::Run() + 311
71:     frame #17: 0x000000010e7afe5a libgtest.dylib`testing::internal::UnitTestImpl::RunAllTests() + 2090
71:     frame #18: 0x000000010e7af488 libgtest.dylib`bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) + 72
71:     frame #19: 0x000000010e7af3f0 libgtest.dylib`testing::UnitTest::Run() + 176
71:     frame #20: 0x000000010e5e106b f_shell`main + 75
71:     frame #21: 0x00007fff5a0f7015 libdyld.dylib`start + 1
71:     frame #22: 0x00007fff5a0f7015 libdyld.dylib`start + 1
71: 
71: " thrown in the test body.
71: [  FAILED  ] atom_diag_real.f_shell_kanamori (4004 ms)
71: [----------] 1 test from atom_diag_real (4004 ms total)
71: 
71: [----------] Global test environment tear-down
71: [==========] 1 test from 1 test case ran. (4004 ms total)
71: [  PASSED  ] 0 tests.
71: [  FAILED  ] 1 test, listed below:
71: [  FAILED  ] atom_diag_real.f_shell_kanamori
71: 
71:  1 FAILED TEST
1/1 Test #71: f_shell ..........................***Failed    4.30 sec
```

**Comment by krivenko** [comment]
@HugoStrand [This patch](https://gist.github.com/krivenko/50652d9c26956936ce274ed9b268f918) solves the problem for me.

**Comment by HugoStrand** [comment]
@krivenko Thank you for the patch! I have collected everything in the pull request #634. Cheers, Hugo

**Comment by parcollet** [comment]
Pending PR review.

**Comment by parcollet** [comment]
Shall we close this ? PR is merged or about to be.

**Comment by HugoStrand** [comment]
merged in #634
